File: Parking_Lot.py
 ___________________                     _______________________
|     Vehicle       |                   |      ParkingSpot       |
|___________________|                   |________________________|
| - licensePlate   |                   | - spotNumber           |
| - vehicleType    |                   | - spotType             |
|__________________|                   | - isAvailable          |
| + Vehicle()      |                   |________________________|
|__________________|                   | + ParkingSpot()       |
                                       |________________________|

                    ___________________
                   |     ParkingLot    |
                   |___________________|
                   | - name            |
                   | - capacity        |
                   | - availableSpots  |
                   | - occupiedSpots   |
                   |___________________|
                   | + parkVehicle()   |
                   | + removeVehicle() |
                   | + findAvailableSpot()|
                   |___________________|


class Vehicle:
    def __init__(self, license_plate, vehicle_type):
        self.license_plate = license_plate
        self.vehicle_type = vehicle_type

class ParkingSpot:
    def __init__(self, spot_number, spot_type, is_available=True):
        self.spot_number = spot_number
        self.spot_type = spot_type
        self.is_available = is_available
        self.vehicle = None  # To keep track of which vehicle is parked

class ParkingLot:
    def __init__(self, name, capacity):
        self.name = name
        self.capacity = capacity
        self.available_spots = []
        self.occupied_spots = []
        self.initialize_spots()

    def initialize_spots(self):
        # Initialize the parking spots (for simplicity, assume half for cars, half for bikes)
        for i in range(1, self.capacity // 2 + 1):
            self.available_spots.append(ParkingSpot(i, 'car'))
        for i in range(self.capacity // 2 + 1, self.capacity + 1):
            self.available_spots.append(ParkingSpot(i, 'bike'))

    def park_vehicle(self, vehicle):
        spot = self.find_available_spot(vehicle.vehicle_type)
        if spot:
            spot.is_available = False
            spot.vehicle = vehicle
            self.occupied_spots.append(spot)
            self.available_spots.remove(spot)
            print(f"Vehicle with license plate {vehicle.license_plate} parked at spot {spot.spot_number}.")
        else:
            print(f"No available spots for vehicle type {vehicle.vehicle_type}.")

    def remove_vehicle(self, vehicle):
        for spot in self.occupied_spots:
            if spot.vehicle.license_plate == vehicle.license_plate:
                spot.is_available = True
                spot.vehicle = None
                self.available_spots.append(spot)
                self.occupied_spots.remove(spot)
                print(f"Vehicle with license plate {vehicle.license_plate} removed from spot {spot.spot_number}.")
                return
        print(f"Vehicle with license plate {vehicle.license_plate} not found in the parking lot.")

    def find_available_spot(self, vehicle_type):
        for spot in self.available_spots:
            if spot.spot_type == vehicle_type and spot.is_available:
                return spot
        return None

# Example usage
if __name__ == "__main__":
    parking_lot = ParkingLot("Main Parking Lot", 10)

    vehicle1 = Vehicle("ABC123", "car")
    vehicle2 = Vehicle("XYZ789", "bike")

    parking_lot.park_vehicle(vehicle1)
    parking_lot.park_vehicle(vehicle2)

    parking_lot.remove_vehicle(vehicle1)
    parking_lot.remove_vehicle(vehicle2)

File: Parking_Lot.java
 ___________________                     _______________________
|     Vehicle       |                   |      ParkingSpot       |
|___________________|                   |________________________|
| - licensePlate   |                   | - spotNumber           |
| - vehicleType    |                   | - spotType             |
|__________________|                   | - isAvailable          |
| + Vehicle()      |                   |________________________|
|__________________|                   | + ParkingSpot()       |
                                       |________________________|

                    ___________________
                   |     ParkingLot    |
                   |___________________|
                   | - name            |
                   | - capacity        |
                   | - availableSpots  |
                   | - occupiedSpots   |
                   |___________________|
                   | + parkVehicle()   |
                   | + removeVehicle() |
                   | + findAvailableSpot()|
                   |___________________|


import java.util.ArrayList;
import java.util.List;

class Vehicle {
    private String licensePlate;
    private String vehicleType;

    public Vehicle(String licensePlate, String vehicleType) {
        this.licensePlate = licensePlate;
        this.vehicleType = vehicleType;
    }

    public String getLicensePlate() {
        return licensePlate;
    }

    public String getVehicleType() {
        return vehicleType;
    }
}

class ParkingSpot {
    private int spotNumber;
    private String spotType;
    private boolean isAvailable;
    private Vehicle parkedVehicle; // To keep track of the parked vehicle

    public ParkingSpot(int spotNumber, String spotType, boolean isAvailable) {
        this.spotNumber = spotNumber;
        this.spotType = spotType;
        this.isAvailable = isAvailable;
        this.parkedVehicle = null;
    }

    public int getSpotNumber() {
        return spotNumber;
    }

    public String getSpotType() {
        return spotType;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    public Vehicle getParkedVehicle() {
        return parkedVehicle;
    }

    public void parkVehicle(Vehicle vehicle) {
        this.isAvailable = false;
        this.parkedVehicle = vehicle;
    }

    public void removeVehicle() {
        this.isAvailable = true;
        this.parkedVehicle = null;
    }
}

class ParkingLot {
    private String name;
    private int capacity;
    private List<ParkingSpot> availableSpots;
    private List<ParkingSpot> occupiedSpots;

    public ParkingLot(String name, int capacity) {
        this.name = name;
        this.capacity = capacity;
        this.availableSpots = new ArrayList<>();
        this.occupiedSpots = new ArrayList<>();
        initializeSpots();
    }

    private void initializeSpots() {
        // Initialize the parking spots (for simplicity, assume half for cars, half for bikes)
        for (int i = 1; i <= capacity / 2; i++) {
            availableSpots.add(new ParkingSpot(i, "car", true));
        }
        for (int i = capacity / 2 + 1; i <= capacity; i++) {
            availableSpots.add(new ParkingSpot(i, "bike", true));
        }
    }

    public void parkVehicle(Vehicle vehicle) {
        ParkingSpot spot = findAvailableSpot(vehicle.getVehicleType());
        if (spot != null) {
            spot.parkVehicle(vehicle);
            occupiedSpots.add(spot);
            availableSpots.remove(spot);
            System.out.println("Vehicle with license plate " + vehicle.getLicensePlate() + " parked at spot " + spot.getSpotNumber() + ".");
        } else {
            System.out.println("No available spots for vehicle type " + vehicle.getVehicleType() + ".");
        }
    }

    public void removeVehicle(Vehicle vehicle) {
        for (ParkingSpot spot : occupiedSpots) {
            if (spot.getParkedVehicle().getLicensePlate().equals(vehicle.getLicensePlate())) {
                spot.removeVehicle();
                availableSpots.add(spot);
                occupiedSpots.remove(spot);
                System.out.println("Vehicle with license plate " + vehicle.getLicensePlate() + " removed from spot " + spot.getSpotNumber() + ".");
                return;
            }
        }
        System.out.println("Vehicle with license plate " + vehicle.getLicensePlate() + " not found in the parking lot.");
    }

    public ParkingSpot findAvailableSpot(String vehicleType) {
        for (ParkingSpot spot : availableSpots) {
            if (spot.getSpotType().equals(vehicleType) && spot.isAvailable()) {
                return spot;
            }
        }
        return null;
    }

    public static void main(String[] args) {
        ParkingLot parkingLot = new ParkingLot("Main Parking Lot", 10);

        Vehicle vehicle1 = new Vehicle("ABC123", "car");
        Vehicle vehicle2 = new Vehicle("XYZ789", "bike");

        parkingLot.parkVehicle(vehicle1);
        parkingLot.parkVehicle(vehicle2);

        parkingLot.removeVehicle(vehicle1);
        parkingLot.removeVehicle(vehicle2);
    }
}
